### PR TITLE
Added nickname templating and associated worker

### PR DESCRIPTION
**Pull request by reddivision**
Short Description: 
Users can now specify a template (using python Formatter syntax) to nickname pokemon in their inventory. The worker for this will process the user's entire inventory each tick, but will only update pokemon that require an update to match the new template (i.e. only caught/hatched pokemon after the first tick).

Use a blank template (`""`) to revert all pokemon to their original names.

Valid names in templates are:
- `name` = pokemon name
- `id` = pokemon type id (e.g. 1 for Bulbasaurs)
- `cp` = pokemon's CP
- `iv_attack` = pokemon's attack IV
- `iv_defense` = pokemon's defense IV
- `iv_stamina` = pokemon's stamina IV
- `iv_ads` = pokemon's IVs in X/X/X format (matches web UI format -- A/D/S)
- `iv_sum` = pokemon's IVs as a sum (e.g. 45 when 3 perfect 15 IVs)
- `iv_pct` = pokemon's IVs as a percentage (0-100)

Examples: 
- `{name}_{iv_pct}` => `Mankey_69`
- `{iv_pct}_{iv_ads}` => `91_15/11/15`
sample

### CODE REVIEW

File: pokemongo_bot/cell_workers/utils.py [hunk 1]
```diff
@@ -144,3 +145,57 @@ def print_yellow(message):
 
 def print_red(message):
     print(u'\033[91m' + message.decode('utf-8') + '\033[0m')
+
+def nickname_pokemon(bot,pokemon):
+    """This requies a pokemon object containing all the standard fields: id, ivs, cp, etc"""
+    template = bot.config.nickname_template.lower()
+    if template == "{name}":
```
Comment (cwild) [inline pokemongo_bot/cell_workers/utils.py]: What if the Pokemon has already been nicknamed? Do you know the existing name at this point?

Comment (reddivision) [inline pokemongo_bot/cell_workers/utils.py]: Except for the case where someone's renaming their entire inventory (which won't be common, I imagine), the pokemon can't have been nicknamed already. In the case where someone wants to "force" a revert to original name, they could use a hack (since I don't trim whitespace) `{name} `. That said, I can refactor all this if there's a desire for it.

To answer your other question, yes, we can know a pokemon is nicknamed by the presence of a `nickname` key in the `pokemon_data` dictionary.

Comment (cwild) [inline pokemongo_bot/cell_workers/utils.py]: Forgive me as I don't know what the pokemon variable looks like at this point, but are you saying ```pokemon.get('nickname')``` exists at this point of the code?

Comment (reddivision) [inline pokemongo_bot/cell_workers/utils.py]: in a manner of speaking, yes. the `pokemon` object may or may not have a `'nickname'` key in it, depending on if it's been nicknamed previously or not.

**Comment by solderzzc** [comment]
This is great, could it also be set to an 'NickName' ?
The usage is:
Eevee evolution is usually random. Unless you change the name of your Eevee to Sparky, Pyro or Rainer. Sparky allows your Eevee to evolve into Jolteon, Pyro allows it to evolve into Flareon and Rainer allows it to evolve into Vaporeon.

**Comment by solderzzc** [comment]
Please also open issue to notify botConfig
https://github.com/PokemonGoF/botConfig

File: configs/config.json.example [hunk 2]
```diff
@@ -35,6 +35,8 @@
     "use_lucky_egg": false,
     "hatch_eggs": true,
     "longer_eggs_first": true,
+    "nickname_template": "{name}_{iv_pct}",
+    "nickname_on_startup": false,
```
Comment (solderzzc) [inline configs/config.json.example]: "nickname_on_startup" was not mentioned in pokecli.py.

**Comment by douglascamata** [comment]
@reddivision please make the nicknaming settings all be embedded inside a key in the config, like the `forts` key embed many other keys. Look for their `add_config` call in `pokecli.py` to learn how to do it.

**Comment by Philipp59** [comment]
recommend to change also the order in catch_worker from

                                iv_display = '{}/{}/{}'.format(
                                    individual_stamina,
                                    individual_attack,
                                    individual_defense

in 
iv_display = '{}/{}/{}'.format(
individual_attack,
individual_defense
individual_stamina,

tested your pull request works like a charm :) 
but i found it quite confusing that your logical and good order from attack/Defense/Stamina didnt match with the catch worker ordering.

also "{name}" in config  resets name to default works also very good.

one problems I noticed.

1. place for letters
There are only 12 letters for nickname so that "{name}""{iv_pct}" doesn't fit for all pokemons 
e.g.:
[17:12:26] Nickname set successfully! Scyther is now Scyther_0.98
[17:12:28] Nickname set successfully! Dratini is now Dratini_0.58
[17:12:30] Nickname set successfully! Electabuzz is now Electabuzz_0

So it would be better to set the default config to "{iv_pct}_{iv_ads}"

Therefore change in utils.py
iv_pct = "{:.2f}".format(iv_sum/45.0) 
to
iv_pct = "{:2f}".format(iv_sum/45.0)*100

with that wie change 0.93 in 93 which is saving 2 letters of space 

to that per default config Pokemon will be names in for e.g. "93_14/14/14" or 100_15/15/15 would fit also

**Comment by Philipp59** [comment]
getting this error now when egg is ready 

```
  File "pokecli.py", line 475, in <module>
    main()
  File "pokecli.py", line 68, in main
    bot.tick()
  File "/Users/XXX/Desktop/bots/PokemonGo-BotFix/pokemongo_bot/__init__.py", line 84, in tick
    if worker.work() == WorkerResult.RUNNING:
  File "/Users/XXX/Desktop/bots/PokemonGo-BotFix/pokemongo_bot/cell_workers/incubate_eggs_worker.py", line 29, in work
    self._hatch_eggs()
  File "/Users/XXX/Desktop/bots/PokemonGo-BotFix/pokemongo_bot/cell_workers/incubate_eggs_worker.py", line 134, in _hatch_eggs
    pokemon['name'] = self.bot.pokemon_list[(pokemon.get('pokemon_id',0)-1)]['Name']
AttributeError: 'NoneType' object has no attribute 'get'
```

Working with 9b2450a7d81b33e0b46cf355c3d4ac9c82e92cfa

**Comment by douglascamata** [comment]
@reddivision if you have the `NicknameAll` behaviour in the beginning of the tick, you don't need to try to set nicknames on pokemons after capture or egg hatches. Also, to make the `NicknameAll` behaviour don't make a request for every single pokemon that was already renamed, you can check the current pokemon name and the new name it would have. If they match, you can skip the pokemon.

**Comment by bigkraig** [comment]
nickname_on_startup doesn't work

File: pokemongo_bot/cell_workers/nickname_pokemon.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+from pokemongo_bot import logger
+from pokemongo_bot.human_behaviour import sleep
+from pokemongo_bot.cell_workers.base_task import BaseTask
+
+class NicknamePokemon(BaseTask):
+    def initialize(self):
+        self.template = self.config.get('nickname_template','').lower().strip()
+        if self.template == "{name}":
+            self.template = ""
```
Comment (reddivision) [inline pokemongo_bot/cell_workers/nickname_pokemon.py]: This might seem unintuitive, but setting the nickname to blank (like in the app) will actually revert the pokemon back to its original name.  So, we set the template to blank to achieve this behavior when we anticipate that is what the user wants.

File: pokemongo_bot/cell_workers/incubate_eggs.py [hunk 4]
```diff
@@ -106,15 +106,13 @@ def _check_inventory(self, lookup_ids=[]):
                         "used": False
                     })
                 elif 'is_egg' not in pokemon and pokemon['id'] in lookup_ids:
-                    matched_pokemon.append({
-                        "pokemon_id": pokemon.get('pokemon_id', -1),
-                        "cp": pokemon.get('cp', 0),
+                    matched_pokemon.append(pokemon.update({
```
Comment (TheSavior) [inline pokemongo_bot/cell_workers/incubate_eggs.py]: Why does a PR that is adding a task for renaming pokemon need to modify the incubate eggs task?

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py [hunk 5]
```diff
@@ -52,9 +50,9 @@ def work(self):
                                 individual_defense = pokemon_data.get("individual_defense", 0)
 
                                 iv_display = '{}/{}/{}'.format(
-                                    individual_stamina,
                                     individual_attack,
-                                    individual_defense
+                                    individual_defense,
```
Comment (TheSavior) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: It seems like the changes to this file aren't relevant either

Comment (Philipp59) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: its a nice feature to resort the Bot ordering of IV from SAD to ADS so its in the same order like the renaming.
It would be quite irritating if a pokemon appears with 4/10/15 (SAD) and then its renamed in 10/15/4 (ADS).
Therefore not change the order in general from SAD (why?) to a more logic on (ADS)

File: pokemongo_bot/cell_workers/incubate_eggs.py (hunk 4 above)
Comment (reddivision) [inline pokemongo_bot/cell_workers/incubate_eggs.py]: It was initially a change required for an earlier iteration of this to work (as it required that pokemon objects passed in maintained their standard keys). As you can see, instead of passing out an object with only a subset of keys, it maintains the original object and just appends new calculated data. I opted to leave it in because I feel this is still a better approach in case the egg incubation worker wants to utilize other keys in the pokemon object. At the very least, it doesn't break anything.

File: pokemongo_bot/cell_workers/pokemon_catch_worker.py (hunk 5 above)
Comment (reddivision) [inline pokemongo_bot/cell_workers/pokemon_catch_worker.py]: The reason that the renaming worker uses ADS format is because that is also used in the web UI. We're using this as an opportunity to make the pokemon catch worker output consistent with this format as well.

File: pokemongo_bot/cell_workers/incubate_eggs.py (hunk 4 above)
Comment (TheSavior) [inline pokemongo_bot/cell_workers/incubate_eggs.py]: If it isn't *needed* by this PR, can you undo these changes? It makes it much scarier to review and accept these PRs since it is hard to tell what the impact across the whole project is going to be. Keeping changes small and focused on a single worker makes our lives way easier. If we need these changes down the road, I'd feel more comfortable making them when it actually impacts something.

File: pokemongo_bot/cell_workers/nickname_pokemon.py (hunk 3 above)
Comment (douglascamata) [inline pokemongo_bot/cell_workers/nickname_pokemon.py]: Please add this behaviour explanation as a comment in the code.

**Comment by douglascamata** [comment]
I'd like to reiterate my suggestion:

> @reddivision if you have the NicknameAll behaviour in the beginning of the tick, you don't need to try to set nicknames on pokemons after capture or egg hatches (they will be renamed in the next tick). Also, to make the NicknameAll behaviour don't make a request for every single pokemon that was already renamed, you can check the current pokemon name and the new name it would have. If they match, you can skip the pokemon.

You can add one config to the worker to make it only rename pokemon on start up, to make it even less expensive, and one config to force rename all pokemon regardless of current name.

**Comment by reddivision** [comment]
@douglascamata your initial suggestion you quoted was that nickname all shouldn't be implemented in catching/hatch methods but should just work as part of the chain (it now does) and that it should check names before submitting request to ensure it doesn't name the same again (it also now does that). Your suggestions here are completely different. To address them: 

What is the use case that someone will want to rename their entire inventory on startup and not after catching/hatching? Is that really in any way practical/useful? The current implementation is that the nicknames of all pokemon will be uniform at all times. 

Also, all pokemon will be renamed regardless of their current name unless it is already the same. With that, what is the point in allowing them to force a rename? It's just going to name it what it already is.

**Comment by douglascamata** [comment]
> What is the use case that someone will want to rename their entire inventory on startup and not after catching/hatching? Is that really in any way practical/useful?

@reddivision when we evolve pokemon, for example, does it handle properly?

**Comment by reddivision** [comment]
@douglascamata yes, it should. the next tick it would recalculate the pokemon's new nickname and it would then be using the pokemon's new id/name in the nickname calculation. it would detect that the nickname does not match the current nickname and fire a request.

**Comment by douglascamata** [comment]
👍 

LGTM

**Comment by TheSavior** [comment]
👍

**Comment by bixuanzju** [comment]
Hi, I have the following config
```
      {
        "type": "NicknamePokemon",
        "config": {
            "nickname_template": "{name}_{iv_ads}"
        }
      },
```
After running, I notice that not every pokemon is renamed as per the config
screen shot 2016-08-01 at 8 24 43 pm
What does `15/` for example mean? The other two are zero?

**Comment by Philipp59** [comment]
change to  **"nickname_template": "{iv_pct}_{iv_ads}"**

**Comment by bixuanzju** [comment]
@Philipp59 yeah, that works, but I want the pokemon's original name plus its ads, could that be done?

**Comment by Philipp59** [comment]
As I wrote above the letter space is limited to 12 so you can't take name+ads.
Why you need the original name of the pokemon ? oO

**Comment by bixuanzju** [comment]
Ah, I see, no particular reason, the current config is fine, thanks!

**Comment by reddivision** [comment]
@bixuanzju check out [the wiki](https://github.com/PokemonGoF/PokemonGo-Bot/wiki/Configuration-files#pokemon-nicknaming) for additional usage information

**Comment by vdubus** [comment]
I would suggest to format the `iv_pct` output on 3 digit so that we can easily sort pokemons by name.

```python
iv_pct = "{:03.0f}".format(100*iv_sum/45.0)
```

Maybe some other template can also gain something from that.

File: configs/config.json.example (hunk 2 above)
Comment (danielsdian) [inline configs/config.json.example]: Any news about renaming only the Eevees?
